include/nektar_interface/particle_cell_mapping/newton_hex.hpp

```diff
@@ -0,0 +1,119 @@
+#ifndef ___NESO_PARTICLE_MAPPING_NEWTON_HEX_H__
+#define ___NESO_PARTICLE_MAPPING_NEWTON_HEX_H__
+
+#include "generated_linear/linear_newton_implementation.hpp"
+#include "particle_cell_mapping_newton.hpp"
+#include <neso_particles.hpp>
+
+using namespace NESO;
+using namespace NESO::Particles;
+
+namespace NESO {
+namespace Newton {
+
+struct MappingHexLinear3D : MappingNewtonIterationBase<MappingHexLinear3D> {
+
+  inline void write_data_v(GeometrySharedPtr geom, void *data_host,
+                           void *data_device) {
+
+    REAL *data_device_real = static_cast<REAL *>(data_device);
+
+    const int num_vertices = 8;
+    NESOASSERT(num_vertices == geom->GetNumVerts(),
+               "Unexpected number of vertices");
+
+    int ix = 0;
+    for (int vx = 0; vx < num_vertices; vx++) {
+      REAL xx[3];
+      auto vertex = geom->GetVertex(vx);
+      vertex->GetCoords(xx[0], xx[1], xx[2]);
+      for (int iy = 0; iy < 3; iy++) {
+        data_device_real[ix + iy] = xx[iy];
+      }
+      ix += 3;
+    }
+
+    // Exit tolerance scaling applied by Nektar++
+    auto m_xmap = geom->GetXmap();
+    auto m_geomFactors = geom->GetGeomFactors();
+    Array<OneD, const NekDouble> Jac =
+        m_geomFactors->GetJac(m_xmap->GetPointsKeys());
+    NekDouble tol_scaling =
+        Vmath::Vsum(Jac.size(), Jac, 1) / ((NekDouble)Jac.size());
+    data_device_real[24] = ABS(1.0 / tol_scaling);
+  }
+
+  inline void free_data_v(void *data_host) { return; }
+
+  inline size_t data_size_host_v() { return 0; }
+
+  inline size_t data_size_device_v() { return (8 * 3 + 1) * sizeof(REAL); }
+
+  inline void newton_step_v(const void *d_data, const REAL xi0, const REAL xi1,
+                            const REAL xi2, const REAL phys0, const REAL phys1,
+                            const REAL phys2, const REAL f0, const REAL f1,
+                            const REAL f2, REAL *xin0, REAL *xin1, REAL *xin2) {
+
+    const REAL *data_device_real = static_cast<const REAL *>(d_data);
+
+    Hexahedron::newton_step_linear_3d(
```

Comment:
probably not worth the boilerplate to splat the vector out into varargs

Reply:
I don't follow?